File: backend/mcp_client.py
"""
MCP Client Module for LangChain Agent.
Connects to external Model Context Protocol (MCP) servers, such as HighByte,
using connection config parameters that create on-demand sessions per tool invocation.
Eliminates long-lived task-group state issues and stream timeouts.
"""

import os
import re
import logging
from datetime import datetime, timedelta, timezone
from typing import List, Tuple, Optional, Dict, Any, AsyncGenerator
from contextlib import asynccontextmanager
from langchain_core.tools import BaseTool

logger = logging.getLogger("mcp_client")


def log_insight_summary(tool_name: str, args: Dict[str, Any], response_result: Any) -> None:
    """
    Logs published insight summaries to insight_logging/ folder.
    """
    try:
        dir_path = os.path.join(os.getcwd(), "insight_logging")
        os.makedirs(dir_path, exist_ok=True)

        now = datetime.now(timezone.utc)
        filename = f"insight_{now.strftime('%Y%m%d_%H%M%S')}_{now.microsecond // 1000}.md"
        filepath = os.path.join(dir_path, filename)

        import json
        args_json = json.dumps(args, indent=2)

        content = (
            f"# Published Insight Summary\n\n"
            f"**Timestamp:** {now.strftime('%Y-%m-%d %H:%M:%S UTC')}\n"
            f"**Tool Name:** {tool_name}\n\n"
            f"## Input Parameters\n```json\n{args_json}\n```\n\n"
            f"## Response Result\n```\n{response_result}\n```\n"
        )

        with open(filepath, "w", encoding="utf-8") as f:
            f.write(content)

        logger.info(f"Insight summary logged successfully to: {filepath}")
    except Exception as e:
        logger.error(f"Failed to log insight summary: {e}")


def fix_insight_payload(insight_val: Any) -> Any:
    """
    Ensures insight parameter is a list of native dictionaries.
    Un-strings stringified JSON arrays/objects passed by LLMs.
    """
    import json
    if isinstance(insight_val, str):
        s = insight_val.strip()
        if (s.startswith("[") and s.endswith("]")) or (s.startswith("{") and s.endswith("}")):
            try:
                parsed = json.loads(s)
                return fix_insight_payload(parsed)
            except Exception:
                return insight_val
        return insight_val

    if isinstance(insight_val, dict):
        return [insight_val]

    if isinstance(insight_val, list):
        cleaned_list = []
        for item in insight_val:
            if isinstance(item, str):
                s = item.strip()
                if (s.startswith("[") and s.endswith("]")) or (s.startswith("{") and s.endswith("}")):
                    try:
                        parsed = json.loads(s)
                        if isinstance(parsed, list):
                            cleaned_list.extend(fix_insight_payload(parsed))
                        elif isinstance(parsed, dict):
                            cleaned_list.append(parsed)
                        else:
                            cleaned_list.append(item)
                    except Exception:
                        cleaned_list.append(item)
                else:
                    cleaned_list.append(item)
            elif isinstance(item, dict):
                cleaned_list.append(item)
            elif isinstance(item, list):
                cleaned_list.extend(fix_insight_payload(item))
            else:
                cleaned_list.append(item)
        return cleaned_list

    return insight_val


def sanitize_mcp_tool_args(tool_name: str, args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Sanitizes and standardizes tool input arguments for HighByte MCP tools.
    Converts relative date/time strings ('now-4h', 'today') into ISO-8601 UTC timestamps.
    Fixes stringified insight JSON payloads into native lists/dictionaries.
    """
    if not isinstance(args, dict):
        return args

    sanitized = dict(args)
    now_utc = datetime.now(timezone.utc)

    # 1. Standardize timestamp arguments for HighByte ISO UTC expectations
    for time_key in ("start_ts", "end_ts", "compare_start_ts", "compare_end_ts"):
        if time_key in sanitized and isinstance(sanitized[time_key], str):
            val = sanitized[time_key].strip()
            val_lower = val.lower()

            if val_lower in ("now", "today"):
                sanitized[time_key] = now_utc.strftime("%Y-%m-%dT%H:%M:%SZ")
            elif val_lower.startswith("now-"):
                match = re.match(r"now-(\d+)([hdm])", val_lower)
                if match:
                    num = int(match.group(1))
                    unit = match.group(2)
                    if unit == "h":
                        st = now_utc - timedelta(hours=num)
                    elif unit == "d":
                        st = now_utc - timedelta(days=num)
                    elif unit == "m":
                        st = now_utc - timedelta(minutes=num)
                    else:
                        st = now_utc - timedelta(hours=4)
                    sanitized[time_key] = st.strftime("%Y-%m-%dT%H:%M:%SZ")
            elif "hour" in val_lower or "day" in val_lower or "ago" in val_lower:
                match = re.search(r"(\d+)\s*(hour|day|min)", val_lower)
                if match:
                    num = int(match.group(1))
                    unit = match.group(2)
                    if "hour" in unit:
                        st = now_utc - timedelta(hours=num)
                    elif "day" in unit:
                        st = now_utc - timedelta(days=num)
                    else:
                        st = now_utc - timedelta(minutes=num)
                    sanitized[time_key] = st.strftime("%Y-%m-%dT%H:%M:%SZ")

    # 2. Enforce agent_name & repair insight payload structure for InsightsPublish
    if "publish" in tool_name.lower() or "insightspublish" in tool_name.lower():
        sanitized["agent_name"] = "J.A.D.A"
        if "insight" in sanitized:
            sanitized["insight"] = fix_insight_payload(sanitized["insight"])

    return sanitized



async def get_highbyte_mcp_tools() -> List[BaseTool]:
    """
    Loads HighByte MCP tools using connection config.
    Each tool invocation executes in its own task-scoped MCP session context,
    preventing stream timeouts and task group cancellation errors.
    """
    enabled_env = os.getenv("HIGHBYTE_MCP_ENABLED", "true").lower()
    if enabled_env not in ("true", "1", "yes"):
        logger.info("HighByte MCP integration is disabled via HIGHBYTE_MCP_ENABLED=false.")
        return []

    mcp_url = os.getenv("HIGHBYTE_MCP_URL", "https://your-mcp-server:8885/mcp")
    bearer_token = os.getenv("HIGHBYTE_MCP_BEARER_TOKEN") or os.getenv("MCP_BEARER_TOKEN")

    if not mcp_url:
        logger.warning("HighByte MCP URL is not configured.")
        return []

    headers = {}
    if bearer_token:
        headers["Authorization"] = f"Bearer {bearer_token}"

    logger.info(f"Loading HighByte MCP tools from: {mcp_url}")

    try:
        from langchain_mcp_adapters.tools import load_mcp_tools

        connection_config = {
            "transport": "streamable_http",
            "url": mcp_url,
            "headers": headers
        }

        mcp_tools = await load_mcp_tools(None, connection=connection_config)
        logger.info(f"Successfully loaded {len(mcp_tools)} HighByte MCP tools via connection config.")
        return mcp_tools
    except Exception as e:
        logger.warning(f"Could not load HighByte MCP tools from {mcp_url}: {e}. Proceeding with local tools only.")
        return []


@asynccontextmanager
async def get_highbyte_mcp_session() -> AsyncGenerator[Tuple[Optional[object], List[BaseTool]], None]:
    """
    Backward-compatible context manager yielding active HighByte MCP tools.
    """
    tools = await get_highbyte_mcp_tools()
    yield None, tools

File: backend_go/pkg/mcp/mcp_client.go
package mcp

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"jada-backend/pkg/formatters"
	"jada-backend/pkg/tools"
)

func LogInsightSummary(toolName string, args map[string]interface{}, responseResult string) {
	dir := "insight_logging"
	if err := os.MkdirAll(dir, 0755); err != nil {
		log.Printf("Failed to create insight_logging directory: %v", err)
		return
	}

	now := time.Now().UTC()
	filename := fmt.Sprintf("insight_%s_%d.md", now.Format("20060102_150405"), now.Nanosecond()/1e6)
	fp := filepath.Join(dir, filename)

	var sb strings.Builder
	sb.WriteString("# Published Insight Summary\n\n")
	sb.WriteString(fmt.Sprintf("**Timestamp:** %s\n", now.Format("2006-01-02 15:04:05 UTC")))
	sb.WriteString(fmt.Sprintf("**Tool Name:** %s\n\n", toolName))

	sb.WriteString("## Input Parameters\n```json\n")
	argsJSON, _ := json.MarshalIndent(args, "", "  ")
	sb.WriteString(string(argsJSON))
	sb.WriteString("\n```\n\n")

	sb.WriteString("## Response Result\n```\n")
	sb.WriteString(responseResult)
	sb.WriteString("\n```\n")

	if err := os.WriteFile(fp, []byte(sb.String()), 0644); err != nil {
		log.Printf("Failed to write insight summary to %s: %v", fp, err)
	} else {
		log.Printf("Insight summary logged successfully to: %s", fp)
	}
}

func fixInsightPayload(v interface{}) interface{} {
	switch val := v.(type) {
	case string:
		strTrim := strings.TrimSpace(val)
		if (strings.HasPrefix(strTrim, "[") && strings.HasSuffix(strTrim, "]")) ||
			(strings.HasPrefix(strTrim, "{") && strings.HasSuffix(strTrim, "}")) {
			var parsed interface{}
			if err := json.Unmarshal([]byte(strTrim), &parsed); err == nil {
				return fixInsightPayload(parsed)
			}
		}
		return val
	case map[string]interface{}:
		return []interface{}{val}
	case []interface{}:
		var cleaned []interface{}
		for _, item := range val {
			subCleaned := fixInsightPayload(item)
			if subList, ok := subCleaned.([]interface{}); ok {
				cleaned = append(cleaned, subList...)
			} else {
				cleaned = append(cleaned, subCleaned)
			}
		}
		return cleaned
	default:
		return v
	}
}

func SanitizeMCPToolArgs(toolName string, args map[string]interface{}) map[string]interface{} {
	if args == nil {
		return args
	}

	sanitized := make(map[string]interface{})
	for k, v := range args {
		sanitized[k] = v
	}

	nowUTC := time.Now().UTC()

	timeKeys := []string{"start_ts", "end_ts", "compare_start_ts", "compare_end_ts"}
	for _, tk := range timeKeys {
		val, ok := sanitized[tk]
		if !ok {
			continue
		}
		valStr, isStr := val.(string)
		if !isStr {
			continue
		}

		valTrimmed := strings.TrimSpace(valStr)
		valLower := strings.ToLower(valTrimmed)

		if valLower == "now" || valLower == "today" {
			sanitized[tk] = nowUTC.Format("2006-01-02T15:04:05Z")
		} else if strings.HasPrefix(valLower, "now-") {
			re := regexp.MustCompile(`now-(\d+)([hdm])`)
			match := re.FindStringSubmatch(valLower)
			if len(match) > 2 {
				num, _ := strconv.Atoi(match[1])
				unit := match[2]
				var st time.Time
				switch unit {
				case "h":
					st = nowUTC.Add(-time.Duration(num) * time.Hour)
				case "d":
					st = nowUTC.Add(-time.Duration(num) * 24 * time.Hour)
				case "m":
					st = nowUTC.Add(-time.Duration(num) * time.Minute)
				default:
					st = nowUTC.Add(-4 * time.Hour)
				}
				sanitized[tk] = st.Format("2006-01-02T15:04:05Z")
			}
		} else if strings.Contains(valLower, "hour") || strings.Contains(valLower, "day") || strings.Contains(valLower, "ago") {
			re := regexp.MustCompile(`(\d+)\s*(hour|day|min)`)
			match := re.FindStringSubmatch(valLower)
			if len(match) > 2 {
				num, _ := strconv.Atoi(match[1])
				unit := match[2]
				var st time.Time
				if strings.Contains(unit, "hour") {
					st = nowUTC.Add(-time.Duration(num) * time.Hour)
				} else if strings.Contains(unit, "day") {
					st = nowUTC.Add(-time.Duration(num) * 24 * time.Hour)
				} else {
					st = nowUTC.Add(-time.Duration(num) * time.Minute)
				}
				sanitized[tk] = st.Format("2006-01-02T15:04:05Z")
			}
		}
	}

	nameLower := strings.ToLower(toolName)
	if strings.Contains(nameLower, "insightspublish") || strings.Contains(nameLower, "publish") {
		sanitized["agent_name"] = "J.A.D.A"
		if rawInsight, ok := sanitized["insight"]; ok {
			sanitized["insight"] = fixInsightPayload(rawInsight)
		}
	}

	return sanitized
}

type MCPToolSchema struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	InputSchema map[string]interface{} `json:"inputSchema"`
}

type MCPListToolsResult struct {
	Tools []MCPToolSchema `json:"tools"`
}

type MCPJSONRPCRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	ID      interface{} `json:"id,omitempty"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
}

type MCPJSONRPCResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      interface{}     `json:"id,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
	} `json:"error,omitempty"`
}

func initMCPSession(mcpURL, bearerToken string) (string, error) {
	client := &http.Client{Timeout: 15 * time.Second}

	initPayload := MCPJSONRPCRequest{
		JSONRPC: "2.0",
		ID:      1,
		Method:  "initialize",
		Params: map[string]interface{}{
			"protocolVersion": "2024-11-05",
			"capabilities":    map[string]interface{}{},
			"clientInfo": map[string]string{
				"name":    "J.A.D.A-Go",
				"version": "1.0.0",
			},
		},
	}

	jsonBytes, err := json.Marshal(initPayload)
	if err != nil {
		return "", err
	}

	req, err := http.NewRequest("POST", mcpURL, bytes.NewBuffer(jsonBytes))
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	if bearerToken != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("MCP initialize returned status %d: %s", resp.StatusCode, string(body))
	}

	sessionID := resp.Header.Get("Mcp-Session-Id")

	// Send initialized notification
	notifPayload := MCPJSONRPCRequest{
		JSONRPC: "2.0",
		Method:  "notifications/initialized",
	}
	notifBytes, _ := json.Marshal(notifPayload)
	notifReq, err := http.NewRequest("POST", mcpURL, bytes.NewBuffer(notifBytes))
	if err == nil {
		notifReq.Header.Set("Content-Type", "application/json")
		if bearerToken != "" {
			notifReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
		}
		if sessionID != "" {
			notifReq.Header.Set("Mcp-Session-Id", sessionID)
		}
		notifResp, err := client.Do(notifReq)
		if err == nil {
			notifResp.Body.Close()
		}
	}

	return sessionID, nil
}

func GetHighByteMCPTools() []tools.ToolDefinition {
	enabledEnv := strings.ToLower(os.Getenv("HIGHBYTE_MCP_ENABLED"))
	if enabledEnv == "false" || enabledEnv == "0" || enabledEnv == "no" {
		log.Println("HighByte MCP integration is disabled via HIGHBYTE_MCP_ENABLED=false.")
		return nil
	}

	mcpURL := os.Getenv("HIGHBYTE_MCP_URL")
	if mcpURL == "" {
		mcpURL = "https://nadefunsdpw01.oshkoshglobal.com:8885/mcp"
	}

	bearerToken := os.Getenv("HIGHBYTE_MCP_BEARER_TOKEN")
	if bearerToken == "" {
		bearerToken = os.Getenv("MCP_BEARER_TOKEN")
	}

	log.Printf("Loading HighByte MCP tools from: %s", mcpURL)

	sessionID, err := initMCPSession(mcpURL, bearerToken)
	if err != nil {
		log.Printf("Could not initialize HighByte MCP session: %v. Proceeding with local tools only.", err)
		return nil
	}

	reqBody := MCPJSONRPCRequest{
		JSONRPC: "2.0",
		ID:      2,
		Method:  "tools/list",
	}

	jsonBytes, err := json.Marshal(reqBody)
	if err != nil {
		log.Printf("Could not marshal MCP list request: %v", err)
		return nil
	}

	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest("POST", mcpURL, bytes.NewBuffer(jsonBytes))
	if err != nil {
		log.Printf("Could not create MCP list request: %v", err)
		return nil
	}

	req.Header.Set("Content-Type", "application/json")
	if bearerToken != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
	}
	if sessionID != "" {
		req.Header.Set("Mcp-Session-Id", sessionID)
	}

	resp, err := client.Do(req)
	if err != nil {
		log.Printf("Could not load HighByte MCP tools from %s: %v. Proceeding with local tools only.", mcpURL, err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		log.Printf("MCP tools/list returned status %d. Proceeding with local tools only.", resp.StatusCode)
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Printf("Could not read MCP response: %v", err)
		return nil
	}

	var jsonResp MCPJSONRPCResponse
	if err := json.Unmarshal(body, &jsonResp); err != nil {
		log.Printf("Could not decode MCP JSON-RPC response: %v", err)
		return nil
	}

	if jsonResp.Error != nil {
		log.Printf("MCP JSON-RPC error: %s", jsonResp.Error.Message)
		return nil
	}

	var listResult MCPListToolsResult
	if err := json.Unmarshal(jsonResp.Result, &listResult); err != nil {
		log.Printf("Could not parse MCP tools list: %v", err)
		return nil
	}

	var mcpTools []tools.ToolDefinition
	for _, t := range listResult.Tools {
		toolName := t.Name
		toolDesc := t.Description
		toolParams := t.InputSchema

		nameLower := strings.ToLower(toolName)
		isInsightPublish := strings.Contains(nameLower, "insightspublish") || strings.Contains(nameLower, "publish")
		isDestructive := strings.Contains(nameLower, "delete") || strings.Contains(nameLower, "remove")

		toolPolicy := tools.ToolPolicy{
			ReadOnly:         !isInsightPublish && !isDestructive,
			Destructive:      isDestructive,
			RequiresApproval: isInsightPublish || isDestructive,
		}

		execFunc := func(args map[string]interface{}) (string, error) {
			sanitizedArgs := SanitizeMCPToolArgs(toolName, args)

			// Execute tool with on-demand MCP session initialization to prevent session expiry
			execSessionID, err := initMCPSession(mcpURL, bearerToken)
			if err != nil {
				execSessionID = sessionID // Fallback to startup sessionID if re-init fails
			}

			callReq := MCPJSONRPCRequest{
				JSONRPC: "2.0",
				ID:      3,
				Method:  "tools/call",
				Params: map[string]interface{}{
					"name":      toolName,
					"arguments": sanitizedArgs,
				},
			}

			callBytes, _ := json.Marshal(callReq)
			httpReq, err := http.NewRequest("POST", mcpURL, bytes.NewBuffer(callBytes))
			if err != nil {
				return fmt.Sprintf("Error executing tool '%s': %v", toolName, err), nil
			}

			httpReq.Header.Set("Content-Type", "application/json")
			if bearerToken != "" {
				httpReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
			}
			if execSessionID != "" {
				httpReq.Header.Set("Mcp-Session-Id", execSessionID)
			}

			callClient := &http.Client{Timeout: 30 * time.Second}
			callResp, err := callClient.Do(httpReq)
			if err != nil {
				return fmt.Sprintf("Error executing tool '%s': %v", toolName, err), nil
			}
			defer callResp.Body.Close()

			callBody, err := io.ReadAll(callResp.Body)
			if err != nil {
				return fmt.Sprintf("Error reading output from tool '%s': %v", toolName, err), nil
			}

			var callJSONResp MCPJSONRPCResponse
			if err := json.Unmarshal(callBody, &callJSONResp); err != nil {
				resStr := string(callBody)
				if isInsightPublish {
					LogInsightSummary(toolName, sanitizedArgs, resStr)
				}
				return formatters.TruncateToolOutput(resStr, formatters.MaxToolOutputChars), nil
			}

			if callJSONResp.Error != nil {
				return fmt.Sprintf("Error executing tool '%s': %s", toolName, callJSONResp.Error.Message), nil
			}

			resStr := string(callJSONResp.Result)
			if isInsightPublish {
				LogInsightSummary(toolName, sanitizedArgs, resStr)
			}
			return formatters.TruncateToolOutput(resStr, formatters.MaxToolOutputChars), nil
		}

		mcpTools = append(mcpTools, tools.ToolDefinition{
			Name:        toolName,
			Description: toolDesc,
			Parameters:  toolParams,
			Policy:      toolPolicy,
			Execute:     execFunc,
		})
	}

	log.Printf("Successfully loaded %d HighByte MCP tools into Go agent.", len(mcpTools))
	return mcpTools
}
